add unmanaged-kcp transport

Molth opened the pull request:
The unmanaged version has better memory utilization, significantly reduces GC, and the speed of processing packets in KCP is two times faster.

Molth commented:
😿Why is no one responding to me? Did I do something wrong? Did I upset you? If so, I am sorry. @miwarnec @MrGadget1024

SoftwareGuy commented:
Got any performance benchmarks to back up these claims? That'll boost possibility of getting it merged into Mirror mainline.

Molth commented:
> Got any performance benchmarks to back up these claims? That'll boost possibility of getting it merged into Mirror mainline.

I wrote a simple benchmark. 
https://github.com/Molth/Benchmark

Molth commented:
public void BenchmarkConnection()
    {
        var bytes = Encoding.UTF8.GetBytes("test");
        stopwatch = Stopwatch.StartNew();
        for (var i = 0; i < 1000; i++)
        {
            client.Send(bytes, KcpChannel.Reliable);
        }

        Console.ForegroundColor = ConsoleColor.Yellow;
        Console.WriteLine(stopwatch.Elapsed);
        Console.ForegroundColor = ConsoleColor.White;
    }

The yellow output in the console shows that managed is always twice as slow as unmanaged.

Molth commented:
I'm not very good at writing benchmarks either, but my friends' projects use the new KCP.
https://github.com/qq362946/Fantasy
https://github.com/egametang/ET

Molth commented:
@SoftwareGuy 
😸Alternatively, could you check this? It’s really fast. 
https://github.com/Molth/asphyxiaUnity

miwarnec commented:
@Molth is this still worth it over threaded C# kcp? and have you compared this against IL2CPP compiled kcp?
perf. should be similar with IL2CPP.

there are a few reasons why we chose pure C# for kcp:
- able to inspect and debug all the code easily
- we fixed a few bugs in kcp that skywind never addressed
- C# version runs on all platforms. native we'll have to compile on windows/mac/linux/mobile/consoles/etc.

James-Frowen commented:
after a quick look, I would guess the increase in performance is from using `memcpy` instead of `MemoryStream`. I would assume there could be similar gain using `byte[]` and [Buffer.BlockCopy](https://learn.microsoft.com/en-us/dotnet/api/system.buffer.blockcopy?view=net-8.0) 

the unsafe/unmanaged code itself shouldn't give any major performance increase over managed c#, because all that will do is skip safety bounds checks on arrays.

Molth commented:
> @Molth is this still worth it over threaded C# kcp? and have you compared this against IL2CPP compiled kcp? perf. should be similar with IL2CPP.
> 
> there are a few reasons why we chose pure C# for kcp:
> 
> * able to inspect and debug all the code easily
> * we fixed a few bugs in kcp that skywind never addressed
> * C# version runs on all platforms. native we'll have to compile on windows/mac/linux/mobile/consoles/etc.

@miwarnec It's still pure C#, and fixed the (bug)[https://github.com/skywind3000/kcp/pull/291]. not native
https://github.com/Molth/Kcp-CSharp

Molth commented:
> after a quick look, I would guess the increase in performance is from using `memcpy` instead of `MemoryStream`. I would assume there could be similar gain using `byte[]` and [Buffer.BlockCopy](https://learn.microsoft.com/en-us/dotnet/api/system.buffer.blockcopy?view=net-8.0)
> 
> the unsafe/unmanaged code itself shouldn't give any major performance increase over managed c#, because all that will do is skip safety bounds checks on arrays.

        private static void memcpy(void* dest, void* src, int n) => Buffer.MemoryCopy(src, dest, n, n);

        private static void memcpy(void* dest, void* src, uint n) => Buffer.MemoryCopy(src, dest, n, n);

Molth commented:
> @Molth is this still worth it over threaded C# kcp? and have you compared this against IL2CPP compiled kcp? perf. should be similar with IL2CPP.
> 
> there are a few reasons why we chose pure C# for kcp:
> 
> * able to inspect and debug all the code easily
> * we fixed a few bugs in kcp that skywind never addressed
> * C# version runs on all platforms. native we'll have to compile on windows/mac/linux/mobile/consoles/etc.

There is basically no difference in socket performance, but kcp processes packets much faster and significantly reduces GC pressure. The final speed is about the same.

Molth commented:
> @Molth is this still worth it over threaded C# kcp? and have you compared this against IL2CPP compiled kcp? perf. should be similar with IL2CPP.
> 
> there are a few reasons why we chose pure C# for kcp:
> 
> * able to inspect and debug all the code easily
> * we fixed a few bugs in kcp that skywind never addressed
> * C# version runs on all platforms. native we'll have to compile on windows/mac/linux/mobile/consoles/etc.

AOT mode, still faster than managed.
